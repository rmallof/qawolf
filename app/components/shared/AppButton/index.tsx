import { Box, Button } from "grommet";
import Link from "next/link";
import styled from "styled-components";

import {
  borderSize,
  colors,
  edgeSize,
  overflowStyle,
  transition,
  transitionDuration,
} from "../../../theme/theme";
import Text from "../Text";
import {
  activeBackground,
  activeBorderColor,
  activeSecondaryBackground,
  background,
  borderColor,
  hoverBackground,
  hoverBorderColor,
  hoverSecondaryBackground,
  textColor,
} from "./config";
import { getBoxPad, getTextMargin, Props, useDisabledStyle } from "./helpers";

function AppButton(props: Props): JSX.Element {
  const {
    IconComponent,
    a11yTitle,
    className,
    color,
    href,
    iconColor,
    iconPosition,
    isDisabled,
    isLarge,
    justify,
    label,
    margin,
    onClick,
    openNewPage,
    type,
  } = props;

  const finalType = useDisabledStyle(props) ? "disabled" : type;

  const innerHtml = (
    <Button
      a11yTitle={a11yTitle || label}
      className={className}
      disabled={isDisabled}
      margin={margin}
      onClick={onClick}
      plain
      style={{ transition }}
    >
      <Box
        align="center"
        direction={iconPosition === "right" ? "row-reverse" : "row"}
        fill
        justify={justify || "between"}
        pad={getBoxPad({
          hasIcon: !!IconComponent,
          hasLabel: !!label,
          iconPosition,
          isLarge,
          justify,
        })}
      >
        {!!IconComponent && (
          <IconComponent
            color={iconColor || color || textColor[finalType]}
            size={edgeSize.small}
          />
        )}
        {!!label && (
          <Text
            color={color || textColor[finalType]}
            margin={getTextMargin(!!IconComponent, iconPosition)}
            size={isLarge ? "buttonLarge" : "component"}
            style={overflowStyle}
          >
            {label}
          </Text>
        )}
      </Box>
    </Button>
  );

  if (href) {
    return (
      <Link href={href}>
        <a target={openNewPage ? "_blank" : undefined}>{innerHtml}</a>
      </Link>
    );
  }

  return innerHtml;
}

const StyledAppButton = styled(AppButton)`
  background: ${(props) =>
    `${background[useDisabledStyle(props) ? "disabled" : props.type]}`};
  border-radius: ${(props) =>
    props.isLarge ? borderSize.medium : borderSize.small};
  height: ${(props) => (props.isLarge ? edgeSize.xxlarge : edgeSize.large)};

  ${(props) => !!props.isDisabled && "cursor: not-allowed;"}
  ${(props) => !!props.width && `width: ${props.width};`}
  border: 1px solid ${(props) =>
    useDisabledStyle(props)
      ? borderColor.disabled
      : borderColor[props.type] || background[props.type]};

  ${(props) =>
    !!props.noBorderSide &&
    `
  border-${props.noBorderSide}: none;
  border-bottom-${props.noBorderSide}-radius: 0;
  border-top-${props.noBorderSide}-radius: 0;
  `}

  ${(props) => !!props.hasError && `border-color: ${colors.danger5};`}
  ${(props) => !!props.isSelected && `border-color: ${colors.gray9};`}

  svg {
    transition: fill ${transitionDuration};
  }

  &:hover {
    ${(props) =>
      !props.isDisabled &&
      `
    background: ${
      hoverSecondaryBackground[props.hoverType] || hoverBackground[props.type]
    };

    border-color: ${
      hoverBorderColor[props.type] ||
      hoverSecondaryBackground[props.hoverType] ||
      hoverBackground[props.type]
    };

    p, 
    svg {
      color: ${textColor[props.hoverType || props.type]};
      fill: ${textColor[props.hoverType || props.type]};
    }
    `}
  }

  &:active {
    ${(props) =>
      !props.isDisabled &&
      `
    background: ${
      activeSecondaryBackground[props.hoverType] || activeBackground[props.type]
    };

    border-color: ${
      activeBorderColor[props.type] ||
      activeSecondaryBackground[props.hoverType] ||
      activeBackground[props.type]
    };
    `}
  }

  ${(props) =>
    props.isDisabled &&
    !useDisabledStyle(props) &&
    `
    opacity: 0.4;
  `}
`;

export default StyledAppButton;
