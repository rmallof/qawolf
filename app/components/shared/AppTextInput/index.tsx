import { Box, BoxProps, TextInput as GrommetTextInput } from "grommet";
import {
  ChangeEvent,
  forwardRef,
  KeyboardEvent,
  Ref,
  useEffect,
  useRef,
  useState,
} from "react";
import styled from "styled-components";

import {
  borderSize,
  colors,
  edgeSize,
  fontFamily,
  fontWeight,
  textDesktop,
  transition,
} from "../../../theme/theme";
import { Size } from "../Text/config";
import ErrorBadge from "./ErrorBadge";

type Props = {
  autoFocus?: boolean;
  error?: string;
  id?: string;
  isDisabled?: boolean;
  isLarge?: boolean;
  margin?: BoxProps["margin"];
  maxLength?: number;
  name?: string;
  onChange: (e: ChangeEvent<HTMLInputElement>) => void;
  onKeyDown?: (e: KeyboardEvent<HTMLInputElement>) => void;
  pad?: { left: string; right: string };
  placeholder?: string;
  size?: Size;
  value: string;
  width?: string;
};

const StyledGrommetTextInput = styled(GrommetTextInput)`
  &:hover {
    border-color: ${colors.gray5} !important;
  }

  &:focus {
    border-color: ${colors.primary} !important;
  }
`;

function TextInput(
  {
    autoFocus,
    error,
    id,
    isDisabled,
    isLarge,
    margin,
    maxLength,
    name,
    onChange,
    onKeyDown,
    pad,
    placeholder,
    size,
    value,
    width,
  }: Props,
  ref?: Ref<HTMLInputElement>
): JSX.Element {
  const errorRef = useRef<HTMLDivElement>(null);
  const [errorWidth, setErrorWidth] = useState(0);

  useEffect(() => {
    setErrorWidth(errorRef.current?.clientWidth || 0);
  }, [error]);

  const finalSize = isLarge ? "buttonLarge" : size || "component";

  const style = {
    borderColor: error ? colors.danger5 : colors.gray3,
    borderRadius: isLarge ? borderSize.medium : borderSize.small,
    borderWidth: borderSize.xsmall,
    color: colors.gray9,
    fontFamily: fontFamily[finalSize],
    fontWeight: fontWeight.normal,
    fontSize: textDesktop[finalSize].size,
    height: isLarge ? edgeSize.xxlarge : edgeSize.large,
    lineHeight: edgeSize.large,
    paddingBottom: 0,
    paddingLeft: isLarge
      ? edgeSize.small
      : pad?.left || `calc(${edgeSize.xsmall} - ${borderSize.xsmall})`,
    paddingRight: isLarge
      ? edgeSize.small
      : pad?.right ||
        `calc(${edgeSize.xsmall} - ${borderSize.xsmall} + ${errorWidth}px)`,
    paddingTop: 0,
    transition,
  };

  return (
    <Box margin={margin} style={{ position: "relative" }} width={width}>
      <StyledGrommetTextInput
        autoFocus={autoFocus}
        disabled={isDisabled}
        id={id}
        maxLength={maxLength}
        name={name}
        onChange={onChange}
        onKeyDown={onKeyDown}
        placeholder={placeholder}
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        ref={ref as any}
        style={style}
        value={value}
      />
      {!!error && <ErrorBadge error={error} ref={errorRef} />}
    </Box>
  );
}

export default forwardRef(TextInput);
