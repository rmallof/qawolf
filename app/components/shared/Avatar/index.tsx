import { Box, Image } from "grommet";

import { edgeSize } from "../../../theme/theme";
import DefaultAvatar from "./DefaultAvatar";

type Props = {
  avatarUrl?: string | null;
  wolfColor: string;
};

export default function Avatar({ avatarUrl, wolfColor }: Props): JSX.Element {
  return (
    <Box
      flex={false}
      height={edgeSize.medium}
      overflow="hidden"
      round="full"
      width={edgeSize.medium}
    >
      {avatarUrl ? (
        <Image fit="contain" src={avatarUrl} />
      ) : (
        <DefaultAvatar color={wolfColor} />
      )}
    </Box>
  );
}
