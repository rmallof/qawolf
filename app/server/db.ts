import knex from "knex";

import environment from "./environment";

export const connectDb = (): knex => {
  const db = knex({
    client: "pg",
    connection: async () => {
      if (environment.NODE_ENV === "test") {
        throw new Error("Use test db");
      }

      const connection: knex.PgConnectionConfig = {
        connectionString: environment.DATABASE_URL,
      };
      if (environment.DATABASE_SSL) connection.ssl = environment.DATABASE_SSL;

      return connection;
    },
  });

  return db;
};
