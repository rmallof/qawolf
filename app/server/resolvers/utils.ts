/* eslint-disable @typescript-eslint/no-non-null-assertion */
import { AuthenticationError } from "../errors";
import { Logger } from "../Logger";
import { findEnvironment } from "../models/environment";
import { findEnvironmentVariable } from "../models/environment_variable";
import { findGroup } from "../models/group";
import { findSuite } from "../models/suite";
import { findTest } from "../models/test";
import { findTrigger } from "../models/trigger";
import { ModelOptions, Team, Test, User } from "../types";

type EnsureEnvironmentAccess = {
  environment_id: string;
  teams: Team[] | null;
};

type EnsureEnvironmentVariableAccess = {
  environment_variable_id: string;
  teams: Team[] | null;
};

type EnsureGroupAccess = {
  group_id: string;
  teams: Team[] | null;
};

type EnsureSuiteAccess = {
  suite_id: string;
  teams: Team[] | null;
};

type EnsureTeams = {
  logger: Logger;
  teams: Team[] | null;
};

type EnsureTeamAccess = {
  logger: Logger;
  team_id: string;
  teams: Team[] | null;
};

type EnsureTestAccess = {
  teams: Team[] | null;
  test?: Test;
  test_id?: string;
};

type EnsureTriggerAccess = {
  teams: Team[] | null;
  trigger_id: string;
};

type EnsureUser = {
  logger: Logger;
  user: User | null;
};

/**
 * @summary Throws if `teams` is falsy
 * @returns `teams`
 */
export const ensureTeams = ({ logger, teams }: EnsureTeams): Team[] => {
  const log = logger.prefix("ensureTeams");

  if (!teams) {
    log.error("teams not provided");
    throw new AuthenticationError("no teams");
  }

  return teams;
};

export const ensureEnvironmentAccess = async (
  { environment_id, teams }: EnsureEnvironmentAccess,
  { db, logger }: ModelOptions
): Promise<Team> => {
  const log = logger.prefix("ensureEnvironmentAccess");

  const teamIds = ensureTeams({ teams, logger }).map((team) => team.id);
  log.debug("ensure teams", teamIds, "can access environment", environment_id);

  const environment = await findEnvironment(environment_id, { db, logger });
  const teamForEnvironment = teams!.find(
    (team) => environment.team_id === team.id
  );

  if (!teamForEnvironment) {
    log.error("teams", teamIds, "cannot access environment", environment_id);
    throw new AuthenticationError("cannot access environment");
  }

  return teamForEnvironment;
};

export const ensureEnvironmentVariableAccess = async (
  { environment_variable_id, teams }: EnsureEnvironmentVariableAccess,
  options: ModelOptions
): Promise<Team> => {
  const log = options.logger.prefix("ensureEnvironmentVariableAccess");

  const teamIds = ensureTeams({ teams, logger: log }).map((team) => team.id);
  log.debug(
    "ensure teams",
    teamIds,
    "can access environment variable",
    environment_variable_id
  );

  const variable = await findEnvironmentVariable(
    environment_variable_id,
    options
  );
  const selectedTeam = teams!.find((team) => variable.team_id === team.id);

  if (!selectedTeam) {
    log.error(
      "teams",
      teamIds,
      "cannot access environment variable",
      environment_variable_id
    );
    throw new AuthenticationError("cannot access environment variable");
  }

  return selectedTeam;
};

export const ensureGroupAccess = async (
  { group_id, teams }: EnsureGroupAccess,
  options: ModelOptions
): Promise<Team> => {
  const log = options.logger.prefix("ensureGroupAccess");

  const teamIds = ensureTeams({ teams, logger: log }).map((team) => team.id);
  log.debug("ensure teams", teamIds, "can access group", group_id);

  const group = await findGroup(group_id, options);
  const selectedTeam = teams!.find((team) => group.team_id === team.id);

  if (!selectedTeam) {
    log.error("teams", teamIds, "cannot access group", group_id);
    throw new AuthenticationError("cannot access group");
  }

  return selectedTeam;
};

/**
 * @summary Throws if `teams` is falsy or doesn't include `suite.team_id`. Can also
 *   throw if `suite_id` doesn't match any suite.
 * @returns The team from `teams` that matches `suite.team_id`, if we didn't throw.
 */
export const ensureSuiteAccess = async (
  { suite_id, teams }: EnsureSuiteAccess,
  options: ModelOptions
): Promise<void> => {
  const log = options.logger.prefix("ensureSuiteAccess");

  const teamIds = ensureTeams({ logger: options.logger, teams }).map(
    (team) => team.id
  );

  log.debug("ensure teams", teamIds, "can access suite", suite_id);

  const suite = await findSuite(suite_id, options);

  if (!teamIds.includes(suite.team_id)) {
    log.error("teams", teamIds, "cannot access suite", suite_id);
    throw new Error("Teams cannot access suite");
  }
};

/**
 * @summary Throws if `teams` is falsy or doesn't include `team_id`
 * @returns The team from `teams` that matches `team_id`, if we didn't throw.
 */
export const ensureTeamAccess = ({
  logger,
  team_id,
  teams,
}: EnsureTeamAccess): Team => {
  const log = logger.prefix("ensureTeamAccess");

  const teamIds = ensureTeams({ logger, teams }).map((team) => team.id);

  const selectedTeam = teams!.find((team) => team.id === team_id);
  if (!selectedTeam) {
    log.error("teams", teamIds, "cannot access team", team_id);
    throw new AuthenticationError("cannot access team");
  }

  log.debug("ensured teams", teamIds, "can access team", team_id);

  return selectedTeam;
};

/**
 * @summary Throws if `teams` is falsy or doesn't include `test.team_id`. Can also
 *   throw if `test_id` doesn't match any test.
 * @returns The team from `teams` that matches `test.team_id`, if we didn't throw.
 */
export const ensureTestAccess = async (
  { teams, test, test_id }: EnsureTestAccess,
  { db, logger }: ModelOptions
): Promise<Team> => {
  const log = logger.prefix("ensureTestAccess");

  const finalTeams = ensureTeams({ logger, teams });
  const teamIds = finalTeams.map((t) => t.id);

  log.debug("ensure teams", teamIds, "can access test", test?.id || test_id);

  const finalTest = test || (await findTest(test_id, { db, logger }));
  const team = finalTeams.find((t) => finalTest.team_id === t.id);

  if (!team) {
    log.error("teams", teamIds, "cannot access test", test?.id || test_id);
    throw new AuthenticationError("cannot access test");
  }

  return team;
};

export const ensureTriggerAccess = async (
  { teams, trigger_id }: EnsureTriggerAccess,
  { db, logger }: ModelOptions
): Promise<Team> => {
  const log = logger.prefix("ensureTriggerAccess");

  const teamIds = ensureTeams({ teams, logger }).map((team) => team.id);
  log.debug("ensure teams", teamIds, "can access trigger", trigger_id);

  const trigger = await findTrigger(trigger_id, { db, logger });
  const selectedTeam = teams!.find((team) => trigger.team_id === team.id);

  if (!selectedTeam) {
    log.error("teams", teamIds, "cannot access trigger", trigger_id);
    throw new AuthenticationError("cannot access trigger");
  }

  return selectedTeam;
};

/**
 * @summary Throws if `user` is falsy
 * @returns `user` if we didn't throw
 */
export const ensureUser = ({ logger, user }: EnsureUser): User => {
  const log = logger.prefix("ensureUser");

  if (!user) {
    log.error("user not provided");
    throw new AuthenticationError("no user");
  }

  return user;
};
