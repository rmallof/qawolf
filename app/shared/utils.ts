// https://stackoverflow.com/a/1214753
export const minutesFromNow = (minutes = 0): string => {
  return new Date(Date.now() + minutes * 60000).toISOString();
};
