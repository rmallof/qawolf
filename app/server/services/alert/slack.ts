import { IncomingWebhook, IncomingWebhookSendArguments } from "@slack/webhook";

import environment from "../../environment";
import { findIntegration } from "../../models/integration";
import { findUsersForTeam } from "../../models/user";
import { ModelOptions, Suite, SuiteRun, Trigger, User } from "../../types";
import { randomChoice } from "../../utils";

type BuildSuiteMessage = {
  runs: SuiteRun[];
  suite: Suite;
  trigger: Trigger | null;
  user: User | null;
};

type PostMessageToSlack = {
  message: IncomingWebhookSendArguments;
  webhook_url: string;
};

type SendSlackAlert = {
  integrationId: string;
  runs: SuiteRun[];
  suite: Suite;
  trigger: Trigger | null;
};

export const buildMessageForSuite = ({
  runs,
  suite,
  trigger,
  user,
}: BuildSuiteMessage): IncomingWebhookSendArguments => {
  const wolfName = user?.wolf_name || "Spirit";
  const wolfVariant = user?.wolf_variant || "white";

  const failingRuns = runs.filter((r) => r.status === "fail");
  const status = failingRuns.length ? "failed." : "passed!";

  const triggerName = trigger?.name || "manually triggered";

  const suiteHref = new URL(`/suites/${suite.id}`, environment.APP_URL).href;
  const headline = `${wolfName} here: <${suiteHref}|${triggerName} tests> ${status}`;
  const text = `${triggerName} tests ${status}`;

  const runBlocks = failingRuns.map((run) => {
    return {
      type: "section",
      text: {
        type: "mrkdwn",
        text: `<${environment.APP_URL}/run/${run.id}|${run.test_name}>`,
      },
      accessory: {
        type: "image",
        image_url: run.gif_url,
        alt_text: run.test_name,
      },
    };
  });

  return {
    blocks: [
      {
        type: "context",
        elements: [
          {
            type: "image",
            image_url: `https://qawolf-public.s3.us-east-2.amazonaws.com/wolf-${wolfVariant}-slack.png`,
            alt_text: wolfName,
          },
          {
            type: "mrkdwn",
            text: headline,
          },
        ],
      },
      ...runBlocks,
    ],
    text,
  };
};

export const postMessageToSlack = async ({
  message,
  webhook_url,
}: PostMessageToSlack): Promise<void> => {
  const webhook = new IncomingWebhook(webhook_url);
  await webhook.send(message);
};

export const sendSlackAlert = async (
  { integrationId, runs, suite, trigger }: SendSlackAlert,
  options: ModelOptions
): Promise<void> => {
  const log = options.logger.prefix("sendSlackAlert");
  log.debug("suite", suite.id);

  try {
    const users = await findUsersForTeam(suite.team_id, options);
    const integration = await findIntegration(integrationId, options);

    if (!integration.webhook_url) {
      log.error("no webhook url for integration", integration.id);
      throw new Error("No webhook url for Slack");
    }

    const user = users.length ? (randomChoice(users) as User) : null;

    await postMessageToSlack({
      message: buildMessageForSuite({ runs, suite, trigger, user }),
      webhook_url: integration.webhook_url,
    });
    log.debug("sent");
  } catch (error) {
    log.alert("error: Slack alert", error.message);
  }
};
