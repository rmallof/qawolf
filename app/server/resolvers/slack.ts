import environment from "../environment";
import { createIntegration } from "../models/integration";
import { updateTeam } from "../models/team";
import { postMessageToSlack } from "../services/alert/slack";
import { createSlackIntegrationUrl, findSlackWebhook } from "../services/slack";
import {
  Context,
  CreateSlackIntegrationMutation,
  CreateUrlMutation,
  Integration,
  SendSlackUpdateMutation,
} from "../types";
import { ensureTeamAccess, ensureUser } from "./utils";

export const createSlackIntegrationResolver = async (
  _: Record<string, unknown>,
  { redirect_uri, slack_code, team_id }: CreateSlackIntegrationMutation,
  { db, logger, teams }: Context
): Promise<Integration> => {
  const log = logger.prefix("createSlackIntegrationResolver");
  const team = ensureTeamAccess({ logger, teams, team_id });

  const slackWebhook = await findSlackWebhook({
    logger,
    redirect_uri,
    slack_code,
  });

  return db.transaction(async (trx) => {
    const integration = await createIntegration(
      {
        settings_url: slackWebhook.settings_url,
        slack_channel: slackWebhook.slack_channel,
        team_id: team.id,
        team_name: slackWebhook.slack_team,
        type: "slack",
        webhook_url: slackWebhook.webhook_url,
      },
      { db: trx, logger }
    );
    log.debug("created integration", integration.id);

    await updateTeam(
      {
        id: team_id,
        is_email_alert_enabled: false,
        alert_integration_id: integration.id,
      },
      { db: trx, logger }
    );
    log.debug("updated team", team_id);

    return integration;
  });
};

export const createSlackIntegrationUrlResolver = (
  _: Record<string, unknown>,
  { redirect_uri }: CreateUrlMutation,
  { logger, user: contextUser }: Context
): string => {
  const log = logger.prefix("createSlackIntegrationUrlResolver");
  const user = ensureUser({ logger, user: contextUser });

  const url = createSlackIntegrationUrl(redirect_uri);
  log.debug("url for user", user.id);

  return url;
};

export const sendSlackUpdateResolver = async (
  _: Record<string, unknown>,
  { message }: SendSlackUpdateMutation,
  { logger, user: contextUser }: Context
): Promise<boolean> => {
  const log = logger.prefix("sendSlackUpdateResolver");
  const user = ensureUser({ logger, user: contextUser });

  log.debug("user", user.id);

  if (environment.SLACK_UPDATES_WEBHOOK) {
    try {
      await postMessageToSlack({
        message: {
          text: `${message} - ${user.email}`,
        },
        webhook_url: environment.SLACK_UPDATES_WEBHOOK,
      });

      return true;
    } catch (error) {
      log.alert("error", error.message);
      return false;
    }
  }

  return false;
};
