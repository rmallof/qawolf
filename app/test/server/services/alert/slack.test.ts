import { findRunsForSuite } from "../../../../server/models/run";
import * as slack from "../../../../server/services/alert/slack";
import * as azure from "../../../../server/services/aws/storage";
import { minutesFromNow } from "../../../../shared/utils";
import { prepareTestDb } from "../../db";
import {
  buildIntegration,
  buildRun,
  buildSuite,
  buildTeam,
  buildTeamUser,
  buildTest,
  buildTrigger,
  buildUser,
  logger,
} from "../../utils";

const { buildMessageForSuite, sendSlackAlert } = slack;

const trigger = buildTrigger({});
const integration = buildIntegration({});
const user = buildUser({});

const db = prepareTestDb();
const options = { db, logger };

beforeAll(async () => {
  await db("users").insert(user);
  await db("teams").insert(buildTeam({}));
  await db("team_users").insert(buildTeamUser({}));

  await db("integrations").insert(integration);
  await db("teams").update({ alert_integration_id: "integrationId" });

  await db("triggers").insert(trigger);

  await db("suites").insert([buildSuite({})]);
  await db("tests").insert(buildTest({ name: "testName" }));
  return db("runs").insert([
    buildRun({
      completed_at: minutesFromNow(),
      status: "fail",
      suite_id: "suiteId",
    }),
    buildRun({
      completed_at: minutesFromNow(),
      i: 2,
      status: "pass",
      suite_id: "suiteId",
    }),
  ]);
});

describe("buildMessageForSuite", () => {
  beforeAll(() => {
    jest.spyOn(azure, "createStorageReadAccessUrl").mockReturnValue("url");
  });

  afterAll(jest.restoreAllMocks);

  it("builds a Slack message for a failing suite", async () => {
    const trigger = await db.select("*").from("triggers").first();
    const suite = await db.select("*").from("suites").first();
    const runs = await findRunsForSuite(suite.id, options);

    expect(
      buildMessageForSuite({ runs, suite, trigger, user })
    ).toMatchSnapshot();
  });

  it("builds a Slack message for a passing suite", async () => {
    await db("runs").where({ id: "runId" }).update({ status: "pass" });

    const trigger = await db.select("*").from("triggers").first();
    const suite = await db.select("*").from("suites").first();
    const runs = await findRunsForSuite(suite.id, options);

    expect(
      buildMessageForSuite({ runs, suite, trigger, user })
    ).toMatchSnapshot();

    expect(
      buildMessageForSuite({ runs, suite, trigger: null, user })
    ).toMatchSnapshot();

    await db("runs").where({ id: "runId" }).update({ status: "fail" });
  });
});

describe("sendSlackAlert", () => {
  let postMessageToSlackSpy: jest.SpyInstance;

  beforeAll(() => {
    postMessageToSlackSpy = jest
      .spyOn(slack, "postMessageToSlack")
      .mockResolvedValue();
  });

  afterEach(jest.clearAllMocks);

  it("sends Slack alert", async () => {
    const trigger = await db.select("*").from("triggers").first();
    const suite = await db.select("*").from("suites").first();
    const runs = await findRunsForSuite(suite.id, options);

    await sendSlackAlert(
      {
        integrationId: "integrationId",
        runs,
        suite,
        trigger,
      },
      options
    );

    expect(postMessageToSlackSpy.mock.calls[0][0]).toEqual({
      message: buildMessageForSuite({ runs, suite, trigger, user }),
      webhook_url: integration.webhook_url,
    });
  });
});
