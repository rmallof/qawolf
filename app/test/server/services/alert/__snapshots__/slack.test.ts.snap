// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildMessageForSuite builds a Slack message for a failing suite 1`] = `
Object {
  "blocks": Array [
    Object {
      "elements": Array [
        Object {
          "alt_text": "Wolf1",
          "image_url": "https://qawolf-public.s3.us-east-2.amazonaws.com/wolf-brown-slack.png",
          "type": "image",
        },
        Object {
          "text": "Wolf1 here: <http://localhost:3000/suites/suiteId|trigger1 tests> failed.",
          "type": "mrkdwn",
        },
      ],
      "type": "context",
    },
    Object {
      "accessory": Object {
        "alt_text": "testName",
        "image_url": "url",
        "type": "image",
      },
      "text": Object {
        "text": "<http://localhost:3000/run/runId|testName>",
        "type": "mrkdwn",
      },
      "type": "section",
    },
  ],
  "text": "trigger1 tests failed.",
}
`;

exports[`buildMessageForSuite builds a Slack message for a passing suite 1`] = `
Object {
  "blocks": Array [
    Object {
      "elements": Array [
        Object {
          "alt_text": "Wolf1",
          "image_url": "https://qawolf-public.s3.us-east-2.amazonaws.com/wolf-brown-slack.png",
          "type": "image",
        },
        Object {
          "text": "Wolf1 here: <http://localhost:3000/suites/suiteId|trigger1 tests> passed!",
          "type": "mrkdwn",
        },
      ],
      "type": "context",
    },
  ],
  "text": "trigger1 tests passed!",
}
`;

exports[`buildMessageForSuite builds a Slack message for a passing suite 2`] = `
Object {
  "blocks": Array [
    Object {
      "elements": Array [
        Object {
          "alt_text": "Wolf1",
          "image_url": "https://qawolf-public.s3.us-east-2.amazonaws.com/wolf-brown-slack.png",
          "type": "image",
        },
        Object {
          "text": "Wolf1 here: <http://localhost:3000/suites/suiteId|manually triggered tests> passed!",
          "type": "mrkdwn",
        },
      ],
      "type": "context",
    },
  ],
  "text": "manually triggered tests passed!",
}
`;
